src/wallet/wallet.h

```diff
@@ -898,13 +899,26 @@ class CWallet : public CCryptoKeyStore, public CValidationInterface
     void ReacceptWalletTransactions();
     void ResendWalletTransactions(int64_t nBestBlockTime, CConnman* connman) override;
     std::vector<uint256> ResendWalletTransactionsBefore(int64_t nTime, CConnman* connman);
+
+    void MarkBalancesDirty();
     CAmount GetBalance() const;
     CAmount GetUnconfirmedBalance() const;
     CAmount GetImmatureBalance() const;
     CAmount GetWatchOnlyBalance() const;
     CAmount GetUnconfirmedWatchOnlyBalance() const;
     CAmount GetImmatureWatchOnlyBalance() const;
 
+    mutable std::atomic<CAmount> nBalanceCache;
+    mutable std::atomic<CAmount> nUnconfirmedBalanceCache;
+    mutable std::atomic<CAmount> nImmatureBalanceCache;
+    mutable std::atomic<CAmount> nWatchOnlyBalanceCache;
+    mutable std::atomic<CAmount> nUnconfirmedWatchOnlyBalanceCache;
+    mutable std::atomic<CAmount> nImmatureWatchOnlyBalanceCache;
+    std::atomic<bool> fBalancesDirty;
+
+    /** Notification for balance changes */
+    boost::signals2::signal<void ()> BalancesDidChange;
```

Comment:
Maybe call it `BalancesChanged` instead of `BalancesDidChange` to be consistent with the naming of other signals (`StatusChanged`, `AddressBookChanged`, etc).